client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java

```diff
@@ -77,26 +88,29 @@
      *                          This is used if the controller endpoints are static and can be directly accessed.
      *                      2. pravega://ip1:port1,ip2:port2,...
      *                          This is used to autodiscovery the controller endpoints from an initial controller list.
+     * @param connectionFactory The factory for connections to the pravega Server
      */
-    public ControllerImpl(final URI controllerURI) {
+    public ControllerImpl(final URI controllerURI, ConnectionFactory connectionFactory) {
```

Comment:
This new parameter here still feels counterintuitive, why would I need to pass a connection factory when instantiating the client-side controller stub? I understand that this is here because we use it in `getRemainingBytes`, but given that this is the only occurrence, then it is possible that `getRemainingBytes` doesn't belong here. I'm actually thinking that perhaps it should be a private method in `ReaderGroupImpl` that is called from `unreadBytes`. One possibility for getting the connection factory is to have a package-protected getter in client factory, which reader group already has.